tls/s2n_client_hello.c

```diff
@@ -433,9 +433,28 @@ int s2n_process_client_hello(struct s2n_connection *conn)
     /* Find potential certificate matches before we choose the cipher. */
     POSIX_GUARD(s2n_conn_find_name_matching_certs(conn));
 
+    /* Save the previous cipher suite */
+    uint8_t previous_cipher_suite_iana[S2N_TLS_CIPHER_SUITE_LEN] = {0, 0};
+    POSIX_CHECKED_MEMCPY(previous_cipher_suite_iana, conn->secure.cipher_suite->iana_value, S2N_TLS_CIPHER_SUITE_LEN);
+
     /* Now choose the ciphers we have certs for. */
     POSIX_GUARD(s2n_set_cipher_as_tls_server(conn, client_hello->cipher_suites.data, client_hello->cipher_suites.size / 2));
 
+    /**
+     *= https://tools.ietf.org/rfc/rfc8446#4.1.4
+     *# Servers MUST ensure that they negotiate the
+     *# same cipher suite when receiving a conformant updated ClientHello (if
+     *# the server selects the cipher suite as the first step in the
+     *# negotiation, then this will happen automatically).
+     **/
+    if (s2n_is_hello_retry_handshake(conn)
+        && previous_cipher_suite_iana[0] != 0
```

Comment:
I was checking for TLS_NULL_WITH_NULL_NULL, which is [0, 0]. I wanted to check for this because I wanted to know if this was the second client hello received or not, in which case the previous client hello wouldn't be TLS_NULL_WITH_NULL_NULL. I made this check more explicit in 0c83506a5f4fa378223d28608192b849b146a060. If I'm understanding your suggestion, I don't think I can check just the cipher_suite because the cipher_suite would have been initialized in either case, just with the null iana value the first time. Is this correct? Or am I missing something?

Reply:
I see what you mean now... The memcmp accidentally fixed it but grep_simple_mistakes.sh complained about it. So this should fix: 569af9b

Reply:
Further fixed in c32170ff786b1bcb824efac9661f7451ce7e63b9